middlewares/retry.go

```diff
@@ -32,12 +33,19 @@ func NewRetry(attempts int, next http.Handler) *Retry {
 }
 
 func (retry *Retry) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	// if we might make multiple attempts, swap the body for an ioutil.NopCloser
+	// cf https://github.com/containous/traefik/issues/1008
+	if retry.attempts > 1 {
+		body := r.Body
+		defer body.Close()
+		r.Body = ioutil.NopCloser(body)
+	}
 	attempts := 1
 	for {
 		recorder := NewRecorder()
 		recorder.responseWriter = rw
 		retry.next.ServeHTTP(recorder, r)
-		if !isNetworkError(recorder.Code) || attempts >= retry.attempts {
+		if recorder.Error == nil || attempts >= retry.attempts {
```

Comment:
Thanks for double checking. The previous filter was looking at response status as it's how the default oxy error handler sets it, but my error handler would pass the error as a new field in the recorder struct. If this field is not nil then it comes from the transport RoundTrip call, meaning it is a network error.

However what I'm not sure about is whether the next attempt would be properly forwarded to the next backend and not the same one, I guess this depends on the load balancer configuration? If it's a sticky session for instance retrying wouldn't really make sense.